File: src/main/source/org/glavo/viewer/util/FontUtils.java
package org.glavo.viewer.util;

import javafx.scene.text.Font;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

public final class FontUtils {
    private FontUtils() {
    }

    public static final String DEFAULT_UI_FONT_FAMILY;
    public static final String DEFAULT_TEXT_FONT_FAMILY;

    static {
        List<String> fonts = Font.getFamilies();
        List<String> uiFontFamilies = Arrays.asList(
                "PingFang SC",
                "Microsoft YaHei UI",
                "Ubuntu",
                "Segoe UI"
        );
        List<String> textFontFamilies = Arrays.asList(
                "Consolas",
                "Source Code Pro",
                "Fira Code",
                "DejaVu Sans Mono",
                "Monospaced"
        );

        String uiFontFamily = null;
        String textFontFamily = null;

        for (String family : uiFontFamilies) {
            if (fonts.contains(family)) {
                uiFontFamily = family;
                break;
            }
        }
        DEFAULT_UI_FONT_FAMILY = uiFontFamily != null ? uiFontFamily : "Dialog";

        for (String family : textFontFamilies) {
            if (fonts.contains(family)) {
                textFontFamily = family;
                break;
            }
        }
        DEFAULT_TEXT_FONT_FAMILY = uiFontFamily != null ? textFontFamily : "Monospaced";
    }
}

File: src/main/kotlin/org/glavo/viewer/util/FontUtils.kt
@file:JvmName("FontUtils")

package org.glavo.viewer.util

import kotlinfx.*

val fontFamily: Regex = """-fx-font-family:\s*"([^"]+)"\s*;""".toRegex()

private val fonts: List<String> = Font.getFamilies()

val defaultUiFontFamily: String =
        fonts.run {
            find { it == "PingFang SC" }
                    ?: find { it == "Microsoft YaHei UI" }
                    ?: find { it == "Ubuntu" }
                    ?: find { it == "Segoe UI" }
                    ?: "Dialog"
        }

val defaultTextFontFamily: String =
        fonts.run {
            find { it == "Consolas" }
                    ?: find { it == "Source Code Pro" }
                    ?: find { it == "Fira Code" }
                    ?: find {it == "DejaVu Sans Mono"}
                    ?: "Monospaced"
        }
